pkg/controller/keyvault/gateway_kv_serviceaccount.go

```diff
@@ -0,0 +1,134 @@
+// workload identity through annotations on their Gateway resources.
+type KvServiceAccountReconciler struct {
+	client client.Client
+	events record.EventRecorder
+}
+
+func NewKvServiceAccountReconciler(mgr ctrl.Manager) error {
+	metrics.InitControllerMetrics(kvSaControllerName)
+
+	return kvSaControllerName.AddToController(ctrl.NewControllerManagedBy(mgr).
+		For(&gatewayv1.Gateway{}), mgr.GetLogger()).
+		Complete(&KvServiceAccountReconciler{
+			client: mgr.GetClient(),
+			events: mgr.GetEventRecorderFor("app-routing-operator"),
+		})
+
+}
+
+func (k *KvServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, retErr error) {
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(kvSaControllerName, res, retErr)
+	}()
+
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating logger from context: %w", err)
+	}
+	logger = kvSaControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	gwObj := &gatewayv1.Gateway{}
+	if err = k.client.Get(ctx, req.NamespacedName, gwObj); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	clientId, err := extractClientIdForManagedSa(gwObj)
+
+	if err != nil {
+		var userErr userError
+		if errors.As(err, &userErr) {
+			logger.Info("user error while extracting clientId from Gateway: %s", userErr.userMessage)
+			k.events.Event(gwObj, corev1.EventTypeWarning, "InvalidInput", userErr.userMessage)
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to extract clientId from Gateway object")
+		return ctrl.Result{}, fmt.Errorf("extracting client ID for app routing serviceaccount: %w", err)
+	}
+
+	if clientId == "" {
+		return ctrl.Result{}, nil
+	}
+
+	toCreate := &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ServiceAccount",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      appRoutingSaName,
+			Namespace: req.Namespace,
+		},
+	}
+
+	logger.Info("checking for existing ServiceAccount")
+
+	existing := &corev1.ServiceAccount{}
+	err = k.client.Get(ctx, types.NamespacedName{Name: toCreate.Name, Namespace: toCreate.Namespace}, existing)
+	if client.IgnoreNotFound(err) != nil {
+		logger.Error(err, "failed to fetch existing app routing serviceaccount")
+		return ctrl.Result{}, fmt.Errorf("checking for existing app routing service account: %w", err)
+	}
+
+	if existing.Annotations != nil && existing.Annotations[wiSaClientIdAnnotation] != "" {
+		existingClientId := existing.Annotations[wiSaClientIdAnnotation]
+		if existingClientId != clientId {
+			errText := fmt.Sprintf("gateway specifies clientId %s but azure-app-routing-kv ServiceAccount already uses clientId %s", clientId, existingClientId)
+			logger.Info(errText)
+			k.events.Event(gwObj, corev1.EventTypeWarning, "InvalidInput", errText)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	toCreate.Annotations = map[string]string{wiSaClientIdAnnotation: clientId}
```

Comment:
why not just put this where you define toCreate?

Reply:
yeah good call that's cleaner